numpy/_core/src/multiarray/unique.cpp

```diff
@@ -34,11 +36,148 @@ FinalAction<F> finally(F f) {
 }
 
 template <typename T>
+size_t hash_integer(const T *value, npy_bool equal_nan) {
+    return std::hash<T>{}(*value);
+}
+
+template <typename S, typename T, S (*real)(T), S (*imag)(T)>
+size_t hash_complex(const T *value, npy_bool equal_nan) {
+    S value_real = real(*value);
+    S value_imag = imag(*value);
+    int hasnan = npy_isnan(value_real) || npy_isnan(value_imag);
+    if (equal_nan && hasnan) {
+        return 0;
+    }
+
+    // Now, equal_nan is false or neither of the values is not NaN.
+    // So we don't need to worry about NaN here.
+    const unsigned char* value_bytes = reinterpret_cast<const unsigned char*>(value);
+    size_t hash = npy_fnv1a(value_bytes, sizeof(T));
+
+    return hash;
+}
+
+size_t hash_complex_clongdouble(const npy_clongdouble *value, npy_bool equal_nan) {
+    npy_longdouble value_real = npy_creall(*value);
+    npy_longdouble value_imag = npy_cimagl(*value);
+    int hasnan = npy_isnan(value_real) || npy_isnan(value_imag);
+    if (equal_nan && hasnan) {
+        return 0;
+    }
+
+    // Now, equal_nan is false or neither of the values is not NaN.
+    // So we don't need to worry about NaN here.
+    // Some floating-point complex dtypes (e.g., npy_complex256) include undefined or
+    // unused bits in their binary representation
+    // (see: https://github.com/numpy/numpy/blob/main/numpy/_core/src/npymath/npy_math_private.h#L254-L261).
+    // Because hashing the raw bit pattern would make the hash depend on those
+    // undefined bits, we extract the mantissa, exponent, and sign components
+    // explicitly and pack them into a buffer to ensure the hash is well-defined.
+    #if defined(HAVE_LDOUBLE_IBM_DOUBLE_DOUBLE_BE) || defined(HAVE_LDOUBLE_IBM_DOUBLE_DOUBLE_LE)
+    constexpr size_t SIZEOF_BUFFER = NPY_SIZEOF_CLONGDOUBLE;
+    const unsigned char* buffer = reinterpret_cast<const unsigned char*>(value);
+    #else
```

Comment:
The approach below looks correct to me. My only comment is that you might be able to add an `#else if` block for platforms where longdouble and double are identical, avoiding the possibly slightly more expensive but standards-compliant version below for 80-bit longdouble on Linux. I don't know offhand if we already have preprocessor macros for platforms where longdouble is double already defined. If not, then this is fine as-is and no need for you to figure that out to merge this PR.